azure_functions_worker/http_v2.py

```diff
@@ -0,0 +1,244 @@
+    @http_request.setter
+    def http_request(self, value):
+        self._http_request = value
+        self._http_request_available_event.set()
+
+    @property
+    def http_response(self):
+        return self._http_response
+
+    @http_response.setter
+    def http_response(self, value):
+        self._http_response = value
+        self._http_response_available_event.set()
+
+    @property
+    def function(self):
+        return self._function
+
+    @function.setter
+    def function(self, value):
+        self._function = value
+
+    @property
+    def fi_context(self):
+        return self._fi_context
+
+    @fi_context.setter
+    def fi_context(self, value):
+        self._fi_context = value
+
+    @property
+    def http_trigger_param_name(self):
+        return self._http_trigger_param_name
+
+    @http_trigger_param_name.setter
+    def http_trigger_param_name(self, value):
+        self._http_trigger_param_name = value
+
+    @property
+    def args(self):
+        return self._args
+
+    @args.setter
+    def args(self, value):
+        self._args = value
+
+    @property
+    def http_request_available_event(self):
+        return self._http_request_available_event
+
+    @property
+    def http_response_available_event(self):
+        return self._http_response_available_event
+
+
+class AsyncContextReference(BaseContextReference):
+    def __init__(self, http_request=None, http_response=None, function=None,
+                 fi_context=None, args=None):
+        super().__init__(event_class=asyncio.Event, http_request=http_request,
+                         http_response=http_response,
+                         function=function, fi_context=fi_context, args=args)
+        self.is_async = True
+
+
+class SingletonMeta(type):
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super().__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class HttpCoordinator(metaclass=SingletonMeta):
+    def __init__(self):
+        self._context_references: Dict[str, BaseContextReference] = {}
+
+    def set_http_request(self, invoc_id, http_request):
+        if invoc_id not in self._context_references:
+            self._context_references[invoc_id] = AsyncContextReference()
+        context_ref = self._context_references.get(invoc_id)
+        context_ref.http_request = http_request
+
+    def set_http_response(self, invoc_id, http_response):
+        if invoc_id not in self._context_references:
+            raise Exception("No context reference found for invocation "
+                            f"{invoc_id}")
+        context_ref = self._context_references.get(invoc_id)
+        context_ref.http_response = http_response
+
+    async def get_http_request_async(self, invoc_id):
+        if invoc_id not in self._context_references:
+            self._context_references[invoc_id] = AsyncContextReference()
+
+        await asyncio.sleep(0)
```

Comment:
Why sleep if its for 0 sec?

Reply:
good catch, its from my experimental code, will remove